client/src/http.rs

```diff
@@ -69,6 +69,59 @@ pub trait RequestBuilder {
     fn body(self, data: Vec<u8>) -> Self;
 }
 
+pub trait RequestManager: Clone {}
+
+#[derive(Clone)]
+pub struct SyncRequestManager<T>
+where
+    T: Clone,
+{
+    pub(crate) inner: T,
+}
+
+impl<T> RequestManager for SyncRequestManager<T>
+where
+    T: SyncSendRequest,
+    T::RequestBuilder: RequestBuilder,
+{
+}
+
+pub trait SyncSendRequest: Clone {
+    type RequestBuilder;
+
+    fn send(&self, rb: Self::RequestBuilder) -> Result<http::Response<Vec<u8>>>;
+}
+
+#[derive(Clone)]
+pub struct AsyncRequestManager<T>
+where
+    T: Clone,
+{
+    inner: T,
+}
+
+impl<T> RequestManager for AsyncRequestManager<T>
+where
+    T: AsyncSendRequest + Clone,
+    T::RequestBuilder: RequestBuilder,
+{
+}
+
+pub trait AsyncSendRequest {
+    type RequestBuilder;
+
+    fn send<'async_trait>(
+        &'async_trait self,
+        rb: Self::RequestBuilder,
+    ) -> ::std::pin::Pin<
+        Box<
+            dyn ::std::future::Future<Output = Result<http::Response<Vec<u8>>>>
+                + Send
+                + 'async_trait,
+        >,
+    >;
+}
```

Comment:
Is this an expanstion of `async_trait` crate?

Reply:
I considered this crate, but it will require users to use this crate as I understand. If we agree with it, then let's use it. 
Btw the native async traits have already stabilized and seem can appear in one of the coming releases of Rust.